File: leet_code_75/my_python_solutions/gcd_of_strings.py
def gcdOfStrings(str1, str2):
    # get the string lengths of each string
    str1_length = len(str1)
    str2_length = len(str2)

    # if theyre the same length
    if str1_length == str2_length:
        # and if theyre the same string, return either or
        if str1 == str2:
            return str1
        # else return the empty string
        else:
            return ''

    # set the final gcd to the empty string
    final_gcd_string = ''

    # if str 1 is smaller
    if str1_length < str2_length:
        # step 1, check if gcd matches with str 2
        # initialize the gcd to be empty
        gcd_string = ''
        # for each index in str 1
        for i in range(str1_length):
            # initialize the concatenated string to be empty
            concatenated_string = ''
            # concatenate the i-th character to the gcd string
            gcd_string += str1[i]
            # for every possible concatentation of gcd upto (str2 length // index + 1)
            for j in range((str2_length // (i + 1))):
                # concatenate gcd to concatenated string
                concatenated_string += gcd_string
            # get the length of the gcd string
            gcd_string_length = len(gcd_string)
            # if the string is a match, and perfectly divides str1 and str2, set the final gcd to the current gcd
            if (concatenated_string == str2) and (str1_length % gcd_string_length == 0) and (str2_length % gcd_string_length == 0):
                final_gcd_string = gcd_string

        # step 2, check if gcd matches with str 1
        # if its the empty string, or if it matches with the str1, return the final gcd
        if final_gcd_string == '' or final_gcd_string == str1:
            return final_gcd_string
        # else check to see if its own cocatenation matches with str 1
        else:
            # get the length of the final gcd string
            final_gcd_string_length = len(final_gcd_string)

            concatenated_string = ''
            # concatenate the final gcd to itself up to (str1 length // final gcd string length)
            for i in range(str1_length // final_gcd_string_length):
                concatenated_string += final_gcd_string 
            # if the concatenation matches with str1, return the final gcd string
            if (concatenated_string == str1):
                return final_gcd_string
            # else return the empty string
            else:
                return ''
    # else str 2 is smaller
    else:
        # step 1, check if gcd matches with str 1
        # initialize the gcd to be empty
        gcd_string = ''
        # for each index in str 2
        for i in range(str2_length):
            # initialize the concatenated string to be empty
            concatenated_string = ''
            # concatenate the i-th character to the gcd string
            gcd_string += str2[i]
            # for every possible concatentation of gcd upto (str1 length // index + 1)
            for j in range((str1_length // (i + 1))):
                # concatenate gcd to concatenated string
                concatenated_string += gcd_string
            # get the length of the gcd string
            gcd_string_length = len(gcd_string)
            # if the string is a match, and perfectly divides str1 and str2, set the final gcd to the current gcd
            if (concatenated_string == str1) and (str1_length % gcd_string_length == 0) and (str2_length % gcd_string_length == 0):
                final_gcd_string = gcd_string

        # step 2, check if gcd matches with str 2
        # if its the empty string, or if it matches with the str2, return the final gcd
        if final_gcd_string == '' or final_gcd_string == str2:
            return final_gcd_string
            # else check to see if its own cocatenation matches with str 2
        else:
            # get the length of the final gcd string
            final_gcd_string_length = len(final_gcd_string)

            concatenated_string = ''
            # concatenate the final gcd to itself up to (str2 length // final gcd string length)
            for i in range(str2_length // final_gcd_string_length):
                concatenated_string += final_gcd_string 
            # if the concatenation matches with str2, return the final gcd string
            if (concatenated_string == str2):
                return final_gcd_string
            # else return the empty string
            else:
                return ''


str1 = 'ABCABC'
str2 = 'ABC'
print("String 1: " + str1 + " String 2: " + str2)
print("------------------------")
print("Result: ABC " + 'Expected: ' + gcdOfStrings(str1, str2) + '\n')

str1 = 'ABABAB'
str2 = 'ABAB'
print("String 1: " + str1 + " String 2: " + str2)
print("------------------------")
print("Result: AB " + 'Expected: ' + gcdOfStrings(str1, str2) + '\n')

str1 = 'ALEX'
str2 = 'ARIAS'
print("String 1: " + str1 + " String 2: " + str2)
print("------------------------")
print("Result: '' " + 'Expected: ' + gcdOfStrings(str1, str2) + '\n')

str1 = 'AAAAAB'
str2 = 'AAA'
print("String 1: " + str1 + " String 2: " + str2)
print("------------------------")
print("Result: '' " + 'Expected: ' + gcdOfStrings(str1, str2) + '\n')

File: leet_code_75/my_java_solutions/GcdOfStrings.java
/** 
    For two strings s and t, we say "t divides s" if and only if s = t + t + t + ... + t + t (i.e., t is concatenated with itself one or more times).
    Given two strings str1 and str2, return the largest string x such that x divides both str1 and str2.

    Example 1:
    Input: str1 = "ABCABC", str2 = "ABC"
    Output: "ABC"

    Example 2:
    Input: str1 = "ABABAB", str2 = "ABAB"
    Output: "AB"

    Example 3:
    Input: str1 = "LEET", str2 = "CODE"
    Output: ""

    Example 4:
    Input: str1 = "AAAAAB", str2 = "AAA"
    Output: ""​​​​​​​

    Constraints:
    1 <= str1.length, str2.length <= 1000
    str1 and str2 consist of English uppercase letters.

*/

public class GcdOfStrings {
    public String gcdOfStrings(String str1, String str2) {
        // extract the length of both strings
        int str1Length = str1.length();
        int str2Length = str2.length();
        // if the string lengths are the same emter
        if (str1Length == str2Length) {
            // the strings are the same
            if (str1.equals(str2)){
                // return either or
                return str1;
                // else return the empty string
            } else {
                return "";
            }
        }

        // declare a gcd of strings
        String gcdOfStrings = "";
        // if string 1 is the smallest enter
        if (str1Length < str2Length) {
            // step 1: check if str2 matches the gcd
            // declare the current gcd
            String gcdString = "";
            // for every i-th character in string 1
            for(int i = 0; i < str1Length; i++){
                // declare a concatenation string
                String concatenatedString = "";
                // set the gcd to the current gcd string + ith character
                gcdString += str1.charAt(i);
                // concatenate the gcd string to itself upto (str2Length / (i + 1) times
                for(int j = 0; j < (str2Length / (i + 1)); j++){
                    concatenatedString += gcdString;
                }
                // get the length of the gcd string
                int gcdStringLength = gcdString.length();
                // if the concatenatedString is the same as string 2 and gcd divides both str1 and str2
                if (concatenatedString.equals(str2) && (str1Length % gcdStringLength == 0) && (str2Length % gcdStringLength == 0)) {
                    // set the gcd of strings to gcd string
                    gcdOfStrings = gcdString;
                }
            }

            // step 2: check if str1 matches the gcd
            // if the strings match, then return gcdOfStrings
            if (gcdOfStrings.equals(str1) || gcdOfStrings.equals("")) {
                return gcdOfStrings;
                // else the strings dont match, so we must check the concatenation
            } else {
                // get the length of the gcd of strings
                int gcdOfStringsLength = gcdOfStrings.length();
                // declare a concatenation string
                String concatenatedString = "";
                // concatenate the gcd of strings to itself upto  str1Length / gcdOfStringsLength
                for (int i = 0; i < str1Length / gcdOfStringsLength; i++) {
                    concatenatedString += gcdOfStrings;
                }
                // if the strings match, then return gcdOfStrings
                if (concatenatedString.equals(str1)) {
                    return gcdOfStrings;
                    // else return the empty string
                } else {
                    return "";
                }
            }
            // else string 2 is the smallest
        } else {
            // step 1: check if str1 matches the gcd
            // declare the current gcd
            String gcdString = "";
            // for every i-th character in string 2
            for(int i = 0; i < str2Length; i++) {
                // declare a concatenation string
                String concatenatedString = "";
                // set the gcd to the current gcd string + ith character
                gcdString += str2.charAt(i);
                // concatenate the gcd string to itself upto (str1Length / (i + 1) times
                for(int j = 0; j < (str1Length / (i + 1)); j++){
                    concatenatedString += gcdString;
                }
                // get the length of the gcd string
                int gcdStringLength = gcdString.length();
                // if the concatenatedString is the same as string 1 and gcd divides both str1 and str2
                if (concatenatedString.equals(str1) && (str1Length % gcdStringLength == 0) && (str2Length % gcdStringLength == 0)) {
                    // set the gcd of strings to gcd string
                    gcdOfStrings = gcdString;
                }
            }

            // step 2: check if str2 matches the gcd
            // if the strings match, then return gcdOfStrings
            if (gcdOfStrings.equals(str2) || gcdOfStrings.equals("")) {
                return gcdOfStrings;
                // else the strings dont match, so we must check the concatenation
            } else {
                // get the length of the gcd of strings
                int gcdOfStringsLength = gcdOfStrings.length();
                // declare a concatenation string
                String concatenatedString = "";
                // concatenate the gcd of strings to itself upto  str2Length / gcdOfStringsLength
                for (int i = 0; i < str2Length / gcdOfStringsLength; i++) {
                    concatenatedString += gcdOfStrings;
                }
                // if the strings match, then return gcdOfStrings
                if (concatenatedString.equals(str2)) {
                    return gcdOfStrings;
                    // else return the empty string
                } else {
                    return "";
                }
            }
        }
    }
    public static void main(String args[]){
        GcdOfStrings gcd = new GcdOfStrings();
        String str1 = "ABCABC";
        String str2 = "ABC";
        System.out.println("String 1: " + str1 + " String 2: " + str2);
        System.out.println("------------------------");
        System.out.println("Result: ABC " + "Expected: " +
        gcd.gcdOfStrings(str1, str2) + "\n");


        str1 = "ABABAB";
        str2 = "ABAB";
        System.out.println("String 1: " + str1 + " String 2: " + str2);
        System.out.println("------------------------");
        System.out.println("Result: AB " + "Expected: " + gcd.gcdOfStrings(str1,
        str2) + "\n");


        str1 = "ALEX";
        str2 = "ARIAS";
        System.out.println("String 1: " + str1 + " String 2: " + str2);
        System.out.println("------------------------");
        System.out.println("Result: '' " + "Expected: " + gcd.gcdOfStrings(str1,
        str2) + "\n");


        str1 = "AAAAAB";
        str2 = "AAA";
        System.out.println("String 1: " + str1 + " String 2: " + str2);
        System.out.println("------------------------");
        System.out.println("Result: '' " + "Expected: " + gcd.gcdOfStrings(str1,
        str2) + "\n");
    }
}
